src/string_pool.rs

```diff
@@ -149,10 +149,13 @@ impl StringPool {
     pub fn intern<'s>(&'s self, s: &str) -> &'s str {
         if s == "" { return ""; }
 
-        let search_string = InternedString::from_str(s);
-
         let mut index = self.index.borrow_mut();
-        let interned_str = *index.entry(search_string).or_insert_with(|| self.do_intern(s));
+        if let Some(interned) = index.get(s) {
+            return unsafe { mem::transmute(interned as &str) };
+        }
+
+        index.insert(self.do_intern(s));
+        let interned_str: &str = index.get(s).unwrap();
```

Comment:
Ideally, there'd be some kind of `entry`-esque API so we reduce it down to a single hash calculation. I think that would need something akin to rust-lang/rfcs#1769 though. 

In the meantime, I think it's OK to have extra hashes for the case where we have to insert anyway. However, can we reduce it from 3 to 2 with something like this?

```rust
let foo = self.do_intern(s);
index.insert(foo);
unsafe { mem::transmute(foo) }
```

That is, it's safe to copy the interned string as the pointer points into the pool and the length is just a number.